R/makeMeta.R

```diff
@@ -0,0 +1,102 @@
+#' Create a metadata object table for a set of charts
+#' 
+#' Generates metadata object for a list charts. makeMeta() looks in chart$package namespace for files called meta_{chart$name} and meta_{chart$domain} for all charts, and then stacks all files. If duplicate metadata rows (domain + text_key) are found an error is thrown. 
+#' 
+#' @param charts list of safetyGraphics chart objects for which to create metadata
+#' 
+#' @return tibble of metadata with the following columns:
+#' \describe{
+#'    \item{domain}{Data domain}
+#'    \item{text_key}{Text key indicating the setting name. \code{'--'} delimiter indicates a field level data mapping}
+#'    \item{col_key}{Key for the column mapping}
+#'    \item{field_key}{Key for the field mapping (if any)}
+#'    \item{type}{type of mapping - "field" or "column"}
+#'    \item{label}{Label}
+#'    \item{description}{Description}
+#'    \item{multiple}{Mapping supports multiple columns/fields }
+#'    \item{standard_adam}{Default values for the ADaM data standard}
+#'    \item{standard_sdtm}{Default values for the SDTM data standard}
+#' }   
+#' 
+#' @importFrom tidyr replace_na starts_with
+#' @export
+
+makeMeta <- function(charts){
+    message(paste0("-Generating meta data for ",length(charts), " charts."))
+    # Check each chart to see if {package}::meta_{domain} or {package}::meta_{name} exists
+    sources <- charts %>% map(function(chart){
+        pkg<-ifelse(is.null(chart$package), 'safetyCharts',chart$package)
+        files<-paste0('meta_',c(chart$name, chart$domain)) %>% map(~list(file=.x, pkg=pkg))
+        return(files)
+    }) %>% 
+    flatten %>%
+    unique
+
+    pkg_dfs<-sources %>% map(function(src){
+        packagePath <- paste0('package:',src$pkg)
+        file_found <- exists(
+            src$file,
+            where=packagePath,
+            inherits=FALSE
+        )
+        if(file_found){
+            this_meta <-  get(
+                src$file,
+                pos=packagePath, 
+                inherits=FALSE
+            )
+
+            if(is.data.frame(this_meta)){
+                this_meta <- this_meta%>%            
+                    mutate(source = paste0(src$pkg, "::",src$file))
+                return(this_meta) 
+            } 
+        }
+    })
+
+    ## check for meta bound directly to the charts
```

Comment:
@jwildfire , What's the workflow to add additional mappings to "standard" meta data? For example, if one wants to add "LBTPT" lab timepoint to meta_labs in a new custom lab data chart, then would need to only add the new mappings and avoid duplicate rows with meta_labs?

Reply:
Yep - that's right, it's fine to add to existing domains, as long as you don't reuse an ID.